Framework/NexusGeometry/src/NexusGeometryParser.cpp

```diff
@@ -595,6 +602,109 @@ void parseMonitors(const H5::Group &root, InstrumentBuilder &builder) {
   }
 }
 
+class Tube {
+private:
+  Eigen::Vector3d axis;
+  double height = 0;
+  double radius;
+  std::vector<Eigen::Vector3d> m_positions;
+  std::vector<int> m_detIDs;
+  Eigen::Vector3d p1;
+  Eigen::Vector3d p2;
+
+public:
+  const Eigen::Vector3d &location() const { return m_positions[0]; }
+  const std::vector<Eigen::Vector3d> &positions() const { return m_positions; }
+  const std::vector<int> &detIDs() const { return m_detIDs; }
+
+  boost::shared_ptr<const Mantid::Geometry::IObject> shape() {
+    Eigen::Matrix<double, 3, 3> points;
+    // Centre shape about (0, 0, 0)
+    auto p1 = m_positions.front();
+    auto p2 = m_positions.back();
+    auto centre = (p1 + p2) / 2;
+
+    points.col(0) = p1 - centre;
+    points.col(2) = p2 - centre;
+
+    // Find point on outer circle
+    auto norm = axis.norm();
+    auto factor = radius / norm;
+    auto pointOnCircle = (axis * factor) + (p1 - centre);
+    points.col(1) = pointOnCircle;
+
+    return NexusShapeFactory::createCylinder(points);
+  }
+
+  Tube(Eigen::Vector3d axis, double radius, Eigen::Vector3d pos, int detID) {
+    this->axis = axis;
+    this->radius = radius;
+    m_positions.push_back(pos);
+    m_detIDs.push_back(detID);
+
+    p1 = axis + pos;
+    p2 = pos;
+  }
+
+  bool addIfCoLinear(const Eigen::Vector3d &pos, int detID) {
+    if (checkCoLinear(pos)) {
+      m_positions.push_back(pos);
+      m_detIDs.push_back(detID);
+
+      auto dist = m_positions[0] - pos;
+      if (dist.norm() > height)
+        height = dist.norm();
+      return true;
+    }
+
+    return false;
+  }
+
+private:
+  bool checkCoLinear(const Eigen::Vector3d &pos) const {
+    auto temp = pos + axis; // unecessary
+    auto numVec = ((p2 - p1).cross(p1 - temp));
+    auto denomVec = (p2 - p1);
+    auto num = numVec.norm();
+
+    if (num == 0)
+      return true;
+
+    auto denom = denomVec.norm();
+    auto d = num / denom;
+
+    return d == 0.0;
+  }
+};
+
+std::vector<Tube> findTubes(const Mantid::Geometry::IObject &shape,
```

Comment:
This important auxiliary function should also have it's own class?/file. Looks like the kind of thing we can and should unit test too.

Reply:
I would say that the following should be put under test and then refactored. Can probably reduce this to much fewer lines.